### PR TITLE
[1.12] Add support for client & server dependencies for mods

### PR DESCRIPTION
**Pull request by mezz**
Based on @Elix-x's work from https://github.com/MinecraftForge/MinecraftForge/pull/2906

This adds support for "client" and "server" dependencies.
For example, this will allow Chisel to depend on the client-side-only CTM mod.
`required-client:ctm;`

~~This also adds support for dependencies that do not need to specify "before" or "after".
This will reduce the number of mod dependency cycles, and is a bit more intuitive.~~

### CODE REVIEW

File: src/test/java/net/minecraftforge/fml/test/DependencyParserTest.java [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+                assertTrue(info.dependants.isEmpty());
+                assertTrue(info.dependencies.isEmpty());
+                assertTrue(info.softRequirements.isEmpty());
+            }
+
+            {
+                String dependencyString = "after:supermod2000@[1.3,);required-before:yetanothermod;softDepMod@[1.0,2.0);required:modW";
+                DependencyParser.DependencyInfo info = parser.parseDependencies(dependencyString);
+                assertContainsSameToString(info.requirements, Sets.newHashSet("yetanothermod", "modW"));
+                assertContainsSameToString(info.softRequirements, Sets.newHashSet("softDepMod@[1.0,2.0)"));
```
Comment (kashike) [inline src/test/java/net/minecraftforge/fml/test/DependencyParserTest.java]: Should use all lowercase, as people may reference this test for examples and think that using uppercase mod ids is okay when it isn't.

**Review by kashike — CHANGES_REQUESTED** [review]

**Comment by Elix-x** [comment]
This looks good to me. Originally, long time ago, i wanted to make a dedicated class, but decided not to, or it would make too much changes, etc, etc... (whatever). I also noticed that you decided to not enforce instructions order, what i only agree with.
Lastly, the dependency of same target can be declared on different sides to use different flags (ex - soft on client, hard on server), which is also nice!

TLDR: 👍

File: src/test/java/net/minecraftforge/fml/test/DependencyParserTest.java (hunk 1 above)
Comment (mezz) [reply src/test/java/net/minecraftforge/fml/test/DependencyParserTest.java]: good catch

Comment (mezz) [reply src/test/java/net/minecraftforge/fml/test/DependencyParserTest.java]: I decided to throw a parsing error when a modId here does not meet the FML requirements, it's too easy to overlook otherwise.

Comment (mezz) [reply src/test/java/net/minecraftforge/fml/test/DependencyParserTest.java]: On second thought, I tried it with some mod packs and they broke horribly. I will make a `TODO` to enable the exception in 1.13.

Comment (marvin-roesch) [reply src/test/java/net/minecraftforge/fml/test/DependencyParserTest.java]: What condition did the mod packs fail on exactly? Anything that runs fine on current 1.11+ should fulfil all the conditions in your sanity check already, shouldn't it?

Comment (mezz) [reply src/test/java/net/minecraftforge/fml/test/DependencyParserTest.java]: I found 6 mods that depended on modIds with capital letters in them. Currently those deps will do nothing, but if I throw an exception it crashes the pack. I reported the issue to those mods and it should be safe to enable this check in 1.13 eince everyone will need to update and test their mods anyway.

Comment (marvin-roesch) [reply src/test/java/net/minecraftforge/fml/test/DependencyParserTest.java]: Hm, in that case, I think you should be at least logging that the dependency string is invalid. People might have hard dependencies on other mods that will end up crashing the game because they're not enforced at load time.

File: src/main/java/net/minecraftforge/fml/common/versioning/DependencyParser.java [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+        {
+            info.dependants.add(artifactVersion);
+        }
+        else if ("after".equals(depOrder))
+        {
+            info.dependencies.add(artifactVersion);
+        }
+        else if (!depRequired && depOrder == null)
+        {
+            info.softRequirements.add(artifactVersion);
```
Comment (marvin-roesch) [inline src/main/java/net/minecraftforge/fml/common/versioning/DependencyParser.java]: This is not strictly speaking a mistake or anything, but if you use the `*` (all) target without any order requirement, you basically have a worthless dependency ("any other mod must be present"). I think we should at least log or document that case.

**Review by marvin-roesch — CHANGES_REQUESTED** [review]

File: src/main/java/net/minecraftforge/fml/common/versioning/DependencyParser.java (hunk 2 above)
Comment (mezz) [reply src/main/java/net/minecraftforge/fml/common/versioning/DependencyParser.java]: That case is already handled, I'll add some test cases for "all" to make sure.

File: src/test/java/net/minecraftforge/fml/test/DependencyParserTest.java (hunk 1 above)
Comment (mezz) [reply src/test/java/net/minecraftforge/fml/test/DependencyParserTest.java]: Makes sense, I'll add logging until 1.13.

File: src/main/java/net/minecraftforge/fml/common/versioning/DependencyParser.java (hunk 2 above)
Comment (marvin-roesch) [reply src/main/java/net/minecraftforge/fml/common/versioning/DependencyParser.java]: Is it? I only gave it a quick glance and didn't find anything that specifically handled a "soft * dependency". Like I said, it's not the end of the world, but users might forget the instructions part and then wonder why their "all" dependency doesn't do anything.

Comment (marvin-roesch) [reply src/main/java/net/minecraftforge/fml/common/versioning/DependencyParser.java]: Nevermind, I found the related code ("Soft dependencies must have a version restriction specified"). Derp xD

Comment (mezz) [reply src/main/java/net/minecraftforge/fml/common/versioning/DependencyParser.java]: In that case it throws "Soft dependencies must have a version restriction specified."
If you give it a version, it throws "You cannot have a versioned dependency on everything (*)"
I have added the unit tests for all the cases now.

File: src/main/java/net/minecraftforge/fml/common/versioning/DependencyParser.java [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+        // TODO: enable this in 1.13
+//        if (!targetIsAll)
+//        {
+//            String modId = artifactVersion.getLabel();
+//            sanityCheckModId(dep, modId);
+//        }
+
+        if (!depRequired && depOrder == null && !targetIsBounded)
+        {
+            throw new DependencyParserException(dep, "Soft dependencies must have a version restriction specified.");
```
Comment (marvin-roesch) [inline src/main/java/net/minecraftforge/fml/common/versioning/DependencyParser.java]: Now that the other thing is out of the way: I am not quite sure what the reasoning behind this restriction would be. To give an example: I might write a Kotlin mod and hence depend on Forgelin. However, I could technically not be using anything special from the Kotlin standard library (except for its auto-generated null-checking stuff, but I'm going to assume that they are pretty much set in stone), so I don't require any particular version of Forgelin/Kotlin, either. Of course I could just specify a version range without an upper bound, but I don't really see a reason for not allowing version-less soft dependencies.

Comment (mezz) [reply src/main/java/net/minecraftforge/fml/common/versioning/DependencyParser.java]: If you need forgelin of any version, you say `"required:forgelin;"`.
If you have a soft dependency on forgelin of any version, you don't need to write any dependency restriction at all, it would do nothing.

Comment (marvin-roesch) [reply src/main/java/net/minecraftforge/fml/common/versioning/DependencyParser.java]: Derp² I shouldn't be looking at this right after getting out of bed xD

**Review by marvin-roesch — APPROVED** [review]

**Comment by mezz** [comment]
Removing support for the "no order" dependencies since it would mess with cpw's much nicer rewrite for 1.13.

**Comment by liach** [comment]
Congratulations!
